services/scheduler/credentials.py

```
@@ -0,0 +1,109 @@
+        Cloned from API's workflow_service.py to maintain compatibility.
+
+        Iterates through all nodes, finds credential references,
+        fetches from database, decrypts, and embeds the values.
+
+        Args:
+            workflow_data: The workflow definition containing nodes and edges
+            db_conn: Database connection for fetching credentials
+
+        Returns:
+            Updated workflow_data with resolved credentials embedded in nodes
+        """
+        # Create a deep copy to avoid mutating the original
+        resolved_data = copy.deepcopy(workflow_data)
+
+        # Get all nodes from the workflow
+        nodes = resolved_data.get("nodes", [])
+
+        logger.info(f"Resolving credentials for {len(nodes)} nodes")
+
+        for node in nodes:
+            # Check if this node has a credentials reference
+            if "credentials" in node and isinstance(node["credentials"], dict):
+                cred_ref = node["credentials"]
+                logger.info(
+                    f"Node {node.get('name')} has credentials reference: {cred_ref}"
+                )
+
+                # If it has an id field, it's a reference that needs resolving
+                if "id" in cred_ref:
+                    credential_id = cred_ref["id"]
+
+                    # Convert string ID to int if needed
+                    if isinstance(credential_id, str):
+                        credential_id = int(credential_id)
+
+                    # Fetch the credential from database
+                    query = """
+                        SELECT id, name, credential_type, credential_data
+                        FROM workflow_credentials
+                        WHERE id = $1
+                    """
+                    credential = await db_conn.fetchrow(query, credential_id)
+
+                    if not credential:
+                        logger.error(f"Credential with ID {credential_id} not found")
+                        continue
```

Silent failure on missing credentials, using continue may leave workflow in invalid state